### PR TITLE
Adding custom testing based on opinfos input for ops with custom rules.

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67575
* __->__ #67500
* #66973
* #66948
* #66898

Differential Revision: [D32497547](https://our.internmc.facebook.com/intern/diff/D32497547)

### CODE REVIEW

File: test/jit/test_dtype_analysis.py [hunk 1]
```diff
@@ -164,10 +177,59 @@ def test_binary_scalar(self):
...
+                input_args = [sample_input.input, *sample_input.args]
+                self.assert_dtype_equal_custom_args(fn, input_args)
+
+        # Now make sure that conv2d doesn't support mixed args
+        conv_ins = sample_inputs_conv2d(None, "cpu", torch.float, False)
+        conv_in = conv_ins[-1]
+        weight, bias = conv_in.args
+        weight = weight.type(torch.long)
+
+        with self.assertRaises(RuntimeError):
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:219]: can have a test to assert that dtype_propagation doesn't dtype the output tensor in this case?

File: test/jit/test_dtype_analysis.py [hunk 2]
```diff
@@ -164,10 +177,59 @@ def test_binary_scalar(self):
...
+            return torch.nn.functional.conv2d(input, weight, bias)
+
+        def adaptive_avg_pool2d_fn(input, output_size: Tuple[int]):
+            return torch._C._nn.adaptive_avg_pool2d(input, output_size)
+
+        for fn, inputs_fn in (
+            (conv2d_fn, sample_inputs_conv2d),
+            (adaptive_avg_pool2d_fn, sample_inputs_adaptive_avg_pool2d),
+        ):
+            for dtype in (torch.int8, torch.float64):
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:207]: how did you decide on these two types?

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: test/jit/test_dtype_analysis.py (hunk 2 above)
Comment (Gamrix) [reply test/jit/test_dtype_analysis.py:207]: Just seemed like two types that would be commonly used, and I wanted one int, and one float type

File: test/jit/test_dtype_analysis.py (hunk 1 above)
Comment (Gamrix) [reply test/jit/test_dtype_analysis.py:219]: Makes sense. I'll have to add a bit more logic to handle that.

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175957).

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175957).

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175957).

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32497547).

**Pull request closed without merging**